ra/ra.go

```diff
@@ -1795,8 +1819,327 @@ func (ra *RegistrationAuthorityImpl) revokeCertificate(ctx context.Context, cert
+			}
+		}
+
+		if features.Enabled(features.MozRevocationReasons) {
+			// Applicants who are not the original Subscriber are not allowed to
+			// revoke for any reason other than cessationOfOperation, which covers
+			// circumstances where "the certificate subscriber no longer owns the
+			// domain names in the certificate". Override the reason code to match.
+			req.Code = ocsp.CessationOfOperation
+			logEvent.Reason = req.Code
+		}
+	}
+
+	issuerID := issuance.GetIssuerNameID(cert)
+	err = ra.revokeCertificate(
+		ctx,
+		cert.SerialNumber,
+		int64(issuerID),
+		revocation.Reason(req.Code),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO(#5979): Check this error when it can't simply be due to a full queue.
+	_ = ra.purgeOCSPCache(ctx, cert, int64(issuerID))
+
+	return &emptypb.Empty{}, nil
+}
+
+// RevokeCertByKey revokes the certificate in question. It always uses
+// reason code 1 (keyCompromise). It ensures that they public key is added to
+// the blocked keys list, even if revocation otherwise fails. It attempts to
+// purge the certificate from the Akamai cache, but it does not hard-fail if
+// doing so is not successful, because the cache will drop the old OCSP response
+// in less than 24 hours anyway.
+func (ra *RegistrationAuthorityImpl) RevokeCertByKey(ctx context.Context, req *rapb.RevokeCertByKeyRequest) (*emptypb.Empty, error) {
+	if req == nil || req.Cert == nil {
+		return nil, errIncompleteGRPCRequest
+	}
+
+	var reason int64
+	if features.Enabled(features.MozRevocationReasons) {
+		// Upcoming Mozilla policy may require that a certificate be revoked with
+		// reason keyCompromise if "the CA obtains verifiable evidence that the
+		// certificate subscriber’s private key corresponding to the public key in
+		// the certificate suffered a key compromise". Signing a JWS to an ACME
+		// server's revocation endpoint certainly counts, so override the reason.
+		reason = ocsp.KeyCompromise
+	} else {
+		if _, present := revocation.UserAllowedReasons[revocation.Reason(req.Code)]; !present {
+			return nil, berrors.BadRevocationReasonError(req.Code)
+		}
+		reason = req.Code
+	}
+
+	cert, err := x509.ParseCertificate(req.Cert)
+	if err != nil {
+		return nil, err
+	}
+
+	issuerID := issuance.GetIssuerNameID(cert)
+
+	logEvent := certificateRevocationEvent{
+		ID:           core.NewToken(),
+		SerialNumber: core.SerialToString(cert.SerialNumber),
+		Reason:       reason,
+		Method:       "key",
+		RequesterID:  0,
+	}
+	defer func() {
+		if err != nil {
+			logEvent.Error = err.Error()
+		}
+		ra.log.AuditObject("Revocation request:", logEvent)
+	}()
+
+	// We revoke the cert before adding it to the blocked keys list, to avoid a
+	// race between this and the bad-key-revoker. But we don't check the error on
+	// from this operation until after we add to the blocked keys list, since that
+	// add needs to happen no matter what.
+	revokeErr := ra.revokeCertificate(
```

Comment:
A consequence of this special error arrangement is that `revokeErr` will never get logged as part of the logEvent. I think it would be good to (a) declare revokeErr earlier to it can be in scope of the defer, and (b) in the defer, if both revokeErr and err are set, concatenate them into the Error field; otherwise pick whichever one is non-nil.

Reply:
Seems like a good approach.